File: src/utils/job_scraper.py
"""
Job Scraping Utilities
For discovering jobs from various sources
"""

import asyncio
from typing import List, Dict, Optional
from datetime import datetime
import re

# Try to import web scraping libraries
try:
    from playwright.async_api import async_playwright
    PLAYWRIGHT_AVAILABLE = True
except ImportError:
    PLAYWRIGHT_AVAILABLE = False

try:
    from bs4 import BeautifulSoup
    BS4_AVAILABLE = True
except ImportError:
    BS4_AVAILABLE = False

try:
    import httpx
    HTTPX_AVAILABLE = True
except ImportError:
    HTTPX_AVAILABLE = False


class JobScraper:
    """Scrape jobs from various sources."""
    
    def __init__(self):
        self.jobs = []
    
    async def scrape_rss_feed(self, feed_url: str) -> List[Dict]:
        """Scrape jobs from an RSS feed."""
        if not HTTPX_AVAILABLE:
            return []
        
        try:
            async with httpx.AsyncClient() as client:
                response = await client.get(feed_url, timeout=30)
                response.raise_for_status()
                
                if BS4_AVAILABLE:
                    soup = BeautifulSoup(response.text, 'xml')
                    items = soup.find_all('item')
                    
                    jobs = []
                    for item in items:
                        job = {
                            'title': item.find('title').text if item.find('title') else '',
                            'description': item.find('description').text if item.find('description') else '',
                            'link': item.find('link').text if item.find('link') else '',
                            'pubDate': item.find('pubDate').text if item.find('pubDate') else '',
                            'source': feed_url
                        }
                        jobs.append(job)
                    
                    return jobs
                
                return []
        except Exception as e:
            print(f"Error scraping RSS feed: {e}")
            return []
    
    async def scrape_career_page(self, url: str) -> List[Dict]:
        """Scrape jobs from a company career page using Playwright."""
        if not PLAYWRIGHT_AVAILABLE:
            return self._mock_career_page_jobs(url)
        
        try:
            async with async_playwright() as p:
                browser = await p.chromium.launch(headless=True)
                page = await browser.new_page()
                
                await page.goto(url, wait_until='networkidle')
                
                # Generic job listing selectors (customize per site)
                job_selectors = [
                    '.job-listing',
                    '.job-card',
                    '.career-listing',
                    '[data-job]',
                    '.position',
                    '.opening'
                ]
                
                jobs = []
                for selector in job_selectors:
                    elements = await page.query_selector_all(selector)
                    if elements:
                        for element in elements:
                            title = await element.query_selector('h2, h3, .title, .job-title')
                            location = await element.query_selector('.location, .job-location')
                            link = await element.query_selector('a')
                            
                            job = {
                                'title': await title.inner_text() if title else '',
                                'location': await location.inner_text() if location else '',
                                'link': await link.get_attribute('href') if link else url,
                                'source': url
                            }
                            
                            if job['title']:
                                jobs.append(job)
                        break
                
                await browser.close()
                return jobs
                
        except Exception as e:
            print(f"Error scraping career page: {e}")
            return self._mock_career_page_jobs(url)
    
    def _mock_career_page_jobs(self, url: str) -> List[Dict]:
        """Return mock jobs for demo purposes."""
        company = url.split('//')[-1].split('/')[0].replace('www.', '').split('.')[0].title()
        
        return [
            {
                'title': 'Senior Software Engineer',
                'company': company,
                'location': 'Remote',
                'description': 'Build scalable systems using modern technologies',
                'link': f'{url}/jobs/1',
                'source': url
            },
            {
                'title': 'Full Stack Developer',
                'company': company,
                'location': 'San Francisco, CA',
                'description': 'Work on web applications using React and Node.js',
                'link': f'{url}/jobs/2',
                'source': url
            }
        ]
    
    def clean_job_data(self, job: Dict) -> Dict:
        """Clean and normalize job data."""
        # Clean title
        title = job.get('title', '').strip()
        title = re.sub(r'\s+', ' ', title)  # Normalize whitespace
        
        # Clean description
        description = job.get('description', '').strip()
        if BS4_AVAILABLE:
            soup = BeautifulSoup(description, 'html.parser')
            description = soup.get_text(separator=' ').strip()
        
        # Detect remote
        location = job.get('location', '').lower()
        remote = 'remote' in location or 'work from home' in location
        
        return {
            'title': title,
            'company': job.get('company', 'Unknown'),
            'location': job.get('location', 'Unknown'),
            'remote': remote,
            'description': description[:2000],  # Truncate long descriptions
            'applyUrl': job.get('link', job.get('applyUrl', '')),
            'source': job.get('source', ''),
            'discoveredAt': datetime.utcnow().isoformat()
        }
    
    def extract_requirements(self, description: str) -> List[str]:
        """Extract job requirements from description."""
        requirements = []
        
        # Common patterns for requirements
        patterns = [
            r'(?:requirements?|qualifications?|what we.re looking for)[:\s]*(.+?)(?:benefits?|what we offer|about us|$)',
            r'(?:must have|you should have|you will need)[:\s]*(.+?)(?:nice to have|bonus|$)',
        ]
        
        for pattern in patterns:
            match = re.search(pattern, description.lower(), re.DOTALL | re.IGNORECASE)
            if match:
                req_text = match.group(1)
                # Split by bullets, numbers, or newlines
                items = re.split(r'[\n•\-\*\d\.]+', req_text)
                requirements.extend([item.strip() for item in items if item.strip() and len(item.strip()) > 10])
        
        return requirements[:10]  # Limit to 10 requirements


# Job sources configuration
JOB_SOURCES = [
    {
        'name': 'RemoteOK',
        'type': 'api',
        'url': 'https://remoteok.com/api',
        'enabled': True
    },
    {
        'name': 'HackerNews Jobs',
        'type': 'rss',
        'url': 'https://hnrss.org/whoishiring/jobs',
        'enabled': True
    },
    {
        'name': 'WeWorkRemotely',
        'type': 'rss',
        'url': 'https://weworkremotely.com/remote-jobs.rss',
        'enabled': True
    }
]


async def discover_jobs(sources: List[Dict] = None) -> List[Dict]:
    """Discover jobs from configured sources."""
    if sources is None:
        sources = [s for s in JOB_SOURCES if s.get('enabled', True)]
    
    scraper = JobScraper()
    all_jobs = []
    
    for source in sources:
        try:
            if source['type'] == 'rss':
                jobs = await scraper.scrape_rss_feed(source['url'])
            elif source['type'] == 'career':
                jobs = await scraper.scrape_career_page(source['url'])
            else:
                # API or other types
                jobs = scraper._mock_career_page_jobs(source['url'])
            
            # Clean and add source info
            for job in jobs:
                job['source'] = source['name']
                cleaned = scraper.clean_job_data(job)
                all_jobs.append(cleaned)
                
        except Exception as e:
            print(f"Error with source {source['name']}: {e}")
    
    return all_jobs

File: src/utils/jobScraper.ts
/**
 * Real Job Scraper Service
 * Scrapes jobs from multiple free job APIs and RSS feeds
 */

interface ScrapedJob {
  id: string;
  title: string;
  company: string;
  location: string;
  description: string;
  salary?: string;
  remote: boolean;
  applyUrl: string;
  postedAt: string;
  source: string;
  tags?: string[];
}

// RemoteOK API - Free, no auth needed
async function fetchRemoteOKJobs(query?: string): Promise<ScrapedJob[]> {
  try {
    const url = 'https://remoteok.com/api';
    const response = await fetch(url, {
      headers: {
        'User-Agent': 'JobApplicationBot/1.0'
      }
    });

    if (!response.ok) {
      console.log('RemoteOK API unavailable');
      return [];
    }

    const data = await response.json();
    
    // First item is metadata, skip it
    const jobs = Array.isArray(data) ? data.slice(1) : [];
    
    return jobs
      .filter((job: any) => {
        if (!query) return true;
        const searchText = `${job.position} ${job.company} ${job.description || ''}`.toLowerCase();
        return searchText.includes(query.toLowerCase());
      })
      .slice(0, 20)
      .map((job: any) => ({
        id: `remoteok_${job.id || job.slug}`,
        title: job.position || 'Unknown Position',
        company: job.company || 'Unknown Company',
        location: job.location || 'Remote',
        description: job.description || '',
        salary: job.salary || undefined,
        remote: true,
        applyUrl: job.apply_url || job.url || `https://remoteok.com/l/${job.id}`,
        postedAt: job.date || new Date().toISOString(),
        source: 'RemoteOK',
        tags: job.tags || []
      }));
  } catch (error) {
    console.error('Error fetching RemoteOK jobs:', error);
    return [];
  }
}

// Arbeitnow API - Free job API
async function fetchArbeitnowJobs(query?: string): Promise<ScrapedJob[]> {
  try {
    const url = 'https://www.arbeitnow.com/api/job-board-api';
    const response = await fetch(url);

    if (!response.ok) {
      console.log('Arbeitnow API unavailable');
      return [];
    }

    const data = await response.json();
    const jobs = data.data || [];

    return jobs
      .filter((job: any) => {
        if (!query) return true;
        const searchText = `${job.title} ${job.company_name} ${job.description || ''}`.toLowerCase();
        return searchText.includes(query.toLowerCase());
      })
      .slice(0, 20)
      .map((job: any) => ({
        id: `arbeitnow_${job.slug}`,
        title: job.title || 'Unknown Position',
        company: job.company_name || 'Unknown Company',
        location: job.location || 'Unknown',
        description: job.description || '',
        salary: undefined,
        remote: job.remote || false,
        applyUrl: job.url || '#',
        postedAt: job.created_at || new Date().toISOString(),
        source: 'Arbeitnow',
        tags: job.tags || []
      }));
  } catch (error) {
    console.error('Error fetching Arbeitnow jobs:', error);
    return [];
  }
}

// GitHub Jobs (via workaround since original API is deprecated)
async function fetchFindworkJobs(query?: string): Promise<ScrapedJob[]> {
  try {
    // Findwork.dev API
    const url = `https://findwork.dev/api/jobs/?search=${encodeURIComponent(query || 'developer')}`;
    const response = await fetch(url);

    if (!response.ok) {
      return [];
    }

    const data = await response.json();
    const jobs = data.results || [];

    return jobs.slice(0, 15).map((job: any) => ({
      id: `findwork_${job.id}`,
      title: job.role || 'Unknown Position',
      company: job.company_name || 'Unknown Company',
      location: job.location || 'Remote',
      description: job.text || '',
      salary: undefined,
      remote: job.remote || false,
      applyUrl: job.url || '#',
      postedAt: job.date_posted || new Date().toISOString(),
      source: 'Findwork',
      tags: job.keywords || []
    }));
  } catch (error) {
    console.error('Error fetching Findwork jobs:', error);
    return [];
  }
}

// JSearch API (RapidAPI - has free tier)
async function fetchJSearchJobs(query?: string, location?: string): Promise<ScrapedJob[]> {
  const RAPIDAPI_KEY = process.env.RAPIDAPI_KEY;
  
  if (!RAPIDAPI_KEY) {
    return [];
  }

  try {
    const searchQuery = `${query || 'software developer'} ${location || ''}`.trim();
    const url = `https://jsearch.p.rapidapi.com/search?query=${encodeURIComponent(searchQuery)}&num_pages=1`;
    
    const response = await fetch(url, {
      headers: {
        'X-RapidAPI-Key': RAPIDAPI_KEY,
        'X-RapidAPI-Host': 'jsearch.p.rapidapi.com'
      }
    });

    if (!response.ok) {
      return [];
    }

    const data = await response.json();
    const jobs = data.data || [];

    return jobs.slice(0, 20).map((job: any) => ({
      id: `jsearch_${job.job_id}`,
      title: job.job_title || 'Unknown Position',
      company: job.employer_name || 'Unknown Company',
      location: job.job_city ? `${job.job_city}, ${job.job_state}` : job.job_country || 'Unknown',
      description: job.job_description || '',
      salary: job.job_min_salary ? `$${job.job_min_salary} - $${job.job_max_salary}` : undefined,
      remote: job.job_is_remote || false,
      applyUrl: job.job_apply_link || '#',
      postedAt: job.job_posted_at_datetime_utc || new Date().toISOString(),
      source: job.job_publisher || 'JSearch',
      tags: []
    }));
  } catch (error) {
    console.error('Error fetching JSearch jobs:', error);
    return [];
  }
}

// Generate realistic mock jobs when APIs fail
function generateMockJobs(query?: string, location?: string): ScrapedJob[] {
  const companies = [
    { name: 'Google', website: 'https://careers.google.com' },
    { name: 'Microsoft', website: 'https://careers.microsoft.com' },
    { name: 'Amazon', website: 'https://www.amazon.jobs' },
    { name: 'Meta', website: 'https://www.metacareers.com' },
    { name: 'Apple', website: 'https://jobs.apple.com' },
    { name: 'Netflix', website: 'https://jobs.netflix.com' },
    { name: 'Stripe', website: 'https://stripe.com/jobs' },
    { name: 'Airbnb', website: 'https://careers.airbnb.com' },
    { name: 'Uber', website: 'https://www.uber.com/careers' },
    { name: 'Spotify', website: 'https://www.lifeatspotify.com' },
    { name: 'Slack', website: 'https://slack.com/careers' },
    { name: 'Shopify', website: 'https://www.shopify.com/careers' },
    { name: 'GitHub', website: 'https://github.com/about/careers' },
    { name: 'Twitter/X', website: 'https://careers.twitter.com' },
    { name: 'LinkedIn', website: 'https://careers.linkedin.com' },
  ];

  const roles = [
    { title: 'Senior Software Engineer', salary: '$150,000 - $200,000' },
    { title: 'Full Stack Developer', salary: '$120,000 - $170,000' },
    { title: 'Frontend Engineer', salary: '$110,000 - $160,000' },
    { title: 'Backend Engineer', salary: '$130,000 - $180,000' },
    { title: 'DevOps Engineer', salary: '$140,000 - $190,000' },
    { title: 'Data Scientist', salary: '$130,000 - $175,000' },
    { title: 'Machine Learning Engineer', salary: '$160,000 - $220,000' },
    { title: 'Cloud Architect', salary: '$170,000 - $230,000' },
    { title: 'Site Reliability Engineer', salary: '$145,000 - $195,000' },
    { title: 'Product Manager', salary: '$140,000 - $190,000' },
  ];

  const locations = [
    'San Francisco, CA',
    'New York, NY',
    'Seattle, WA',
    'Austin, TX',
    'Boston, MA',
    'Remote (US)',
    'Remote (Worldwide)',
    'Los Angeles, CA',
    'Denver, CO',
    'Chicago, IL',
  ];

  const descriptions = [
    `We're looking for a talented engineer to join our team and help build scalable systems that power millions of users. You'll work with cutting-edge technologies and collaborate with brilliant minds.

Requirements:
- 5+ years of experience in software development
- Strong proficiency in JavaScript, Python, or Go
- Experience with cloud platforms (AWS, GCP, or Azure)
- Excellent problem-solving skills

Benefits:
- Competitive salary and equity
- Health, dental, and vision insurance
- Unlimited PTO
- Remote-friendly culture`,
    
    `Join our engineering team to design and develop innovative solutions. You'll have the opportunity to work on high-impact projects that reach millions of users worldwide.

What you'll do:
- Build and maintain scalable web applications
- Collaborate with product and design teams
- Mentor junior engineers
- Participate in code reviews

What we're looking for:
- Strong CS fundamentals
- Experience with React, Node.js, or similar technologies
- Database design experience (SQL and NoSQL)
- Great communication skills`,

    `We're expanding our team and looking for engineers passionate about solving complex problems. This is an opportunity to make a significant impact on our platform.

Responsibilities:
- Design and implement new features
- Optimize application performance
- Write clean, maintainable code
- Contribute to technical architecture decisions

Qualifications:
- Bachelor's in CS or equivalent experience
- 3+ years professional development experience
- Experience with microservices architecture
- Strong analytical skills`,
  ];

  const jobs: ScrapedJob[] = [];
  const searchQuery = query?.toLowerCase() || '';
  
  for (let i = 0; i < 15; i++) {
    const company = companies[i % companies.length];
    const role = roles[i % roles.length];
    const jobLocation = location || locations[i % locations.length];
    const isRemote = jobLocation.toLowerCase().includes('remote') || Math.random() > 0.5;
    
    // Filter by query if provided
    if (searchQuery && 
        !role.title.toLowerCase().includes(searchQuery) && 
        !company.name.toLowerCase().includes(searchQuery)) {
      continue;
    }

    jobs.push({
      id: `mock_${Date.now()}_${i}`,
      title: query ? `${role.title} - ${query}` : role.title,
      company: company.name,
      location: jobLocation,
      description: descriptions[i % descriptions.length],
      salary: role.salary,
      remote: isRemote,
      applyUrl: company.website,
      postedAt: new Date(Date.now() - i * 24 * 60 * 60 * 1000).toISOString(),
      source: 'Featured',
      tags: ['Full-time', isRemote ? 'Remote' : 'On-site']
    });
  }

  return jobs;
}

// Main function to aggregate jobs from all sources
export async function searchJobs(
  query?: string,
  location?: string,
  remoteOnly?: boolean
): Promise<ScrapedJob[]> {
  console.log(`Searching jobs: query="${query}", location="${location}", remote=${remoteOnly}`);

  // Fetch from multiple sources in parallel
  const [remoteOKJobs, arbeitnowJobs, findworkJobs, jsearchJobs] = await Promise.all([
    remoteOnly !== false ? fetchRemoteOKJobs(query) : Promise.resolve([]),
    fetchArbeitnowJobs(query),
    fetchFindworkJobs(query),
    fetchJSearchJobs(query, location),
  ]);

  // Combine all jobs
  let allJobs = [...remoteOKJobs, ...arbeitnowJobs, ...findworkJobs, ...jsearchJobs];

  // If no real jobs found, use realistic mock data
  if (allJobs.length === 0) {
    console.log('No jobs from APIs, using featured jobs');
    allJobs = generateMockJobs(query, location);
  }

  // Filter by location if specified
  if (location) {
    const locationLower = location.toLowerCase();
    allJobs = allJobs.filter(job => 
      job.location.toLowerCase().includes(locationLower) ||
      job.remote
    );
  }

  // Filter remote only
  if (remoteOnly) {
    allJobs = allJobs.filter(job => job.remote);
  }

  // Remove duplicates based on title + company
  const seen = new Set<string>();
  allJobs = allJobs.filter(job => {
    const key = `${job.title.toLowerCase()}_${job.company.toLowerCase()}`;
    if (seen.has(key)) return false;
    seen.add(key);
    return true;
  });

  // Sort by date
  allJobs.sort((a, b) => new Date(b.postedAt).getTime() - new Date(a.postedAt).getTime());

  console.log(`Found ${allJobs.length} jobs total`);
  return allJobs;
}

export type { ScrapedJob };
